### PR TITLE
Vendoring libnetwork 83743db8ceb2bdbfa0960d9a54ed2f98df4ea846

### PR DESCRIPTION
**Pull request by mavenugo**
This PR brings in 
- Native multi-host Networking using overlay tunneling (vxlan to begin with)
- Daemon flag --kv-store to configure libkv provider required for multi-host distributed states.
- Adjusted ReleaseNetwork to the new CNM API that provides LeaveAll API.

Native multihost networking can be properly exercised only under Experimental channel as all the
relevant UI `docker network`, `docker service` and `--publish-service` UI are all under experimental tag.

### CODE REVIEW

**Comment by estesp** [comment]
FYI @tiborvass @jfrazelle @jhowardmsft @vbatts and others who've been wondering/looking..this fixes the issue I found last night that was causing all the CI failures in docker-py `TestRemoveLink`.  Specifically, there was a `defer` ordering problem in libnetwork's `endpoint.go` which meant it was many times dereferencing a nil pointer and crashing the daemon, causing the docker-py tests to timeout and fail.

**Comment by vbatts** [comment]
Phil, I owe a big ole hug

**Comment by icecrime** [comment]
@mavenugo The PR has no docs, can you please tell us more about `--kv-store` (or add the docs if possible)? Also I'm surprised it's not experimental: does it mean that multi-host networking is ready for mainstream, but not the `docker network`/`docker service` UX just yet?

**Comment by mavenugo** [comment]
@icecrime good point on experimental for --kv-store. I will fix that.
Since we are controlling the UX under experimental, the multi-host networking cannot be exercised with the existing non-experimental version.

File: daemon/container_linux.go [hunk 1]
```diff
@@ -793,12 +807,29 @@ func (container *Container) AllocateNetwork() error {
...
 
+	if err := container.configureNetwork(networkName, service, networkDriver, mode.IsDefault()); err != nil {
+		return err
+	}
+
+	if err := container.WriteHostConfig(); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (runcom) [inline daemon/container_linux.go]: nit, this could be `return container.WriteHostConfig()` w/o the previous `if`

**Comment by mavenugo** [comment]
@icecrime @runcom addressed your comments. Moved the flags `--default-network` and `--kv-store` under experimental flag. and wrote appropriate docs.
Also added tests for --default-network. But tests for `--kv-store` requires a richer integration framework for multi-host networking and it will be taken up soon in subsequent PRs.

**Comment by moxiegirl** [comment]
LGTM

**Comment by icecrime** [comment]
LGTM

**Comment by calavera** [comment]
LGTM

**Pull request merged**